### PR TITLE
[RFC] Restructure package

### PR DESCRIPTION
**Pull request by OrangeChannel**
One possible solution to https://github.com/Irrational-Encoding-Wizardry/vsutil/issues/38.

```py
>>> import vsutil
>>> [x for x in dir(vsutil) if not x.startswith('_')]
['Dither',
 'Range',
 'decorators',
 'depth',
 'disallow_variable_format',
 'disallow_variable_resolution',
 'enums',
 'fallback',
 'frame2clip',
 'get_depth',
 'get_plane_size',
 'get_subsampling',
 'get_w',
 'get_y',
 'insert_clip',
 'is_image',
 'iterate',
 'join',
 'main',
 'plane',
 'scale_value',
 'should_dither',
 'split']

>>> vsutil.__all__
['disallow_variable_format',
 'disallow_variable_resolution',
 'Dither',
 'Range',
 'fallback',
 'get_w',
 'is_image',
 'iterate',
 'scale_value',
 'should_dither',
 'get_depth',
 'get_plane_size',
 'get_subsampling',
 'depth',
 'frame2clip',
 'get_y',
 'insert_clip',
 'join',
 'plane',
 'split']
```

```py
>>> default = dir()
>>> from vsutil import *
>>> [var for var in dir() if var not in default]
['Dither',
 'Range',
 '_i3',  # iPython
 '_i4',  # iPython
 'default',  # above
 'depth',
 'disallow_variable_format',
 'disallow_variable_resolution',
 'fallback',
 'frame2clip',
 'get_depth',
 'get_plane_size',
 'get_subsampling',
 'get_w',
 'get_y',
 'insert_clip',
 'is_image',
 'iterate',
 'join',
 'plane',
 'scale_value',
 'should_dither',
 'split']
```

This makes it so wildcard imports act exactly as they do now (i.e. `Range`), while also having the ability to use `vsutil.enums.Range` if you import the package normally.

### CODE REVIEW

File: vsutil/enums.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                 var_name: str,
+                 module: Optional[str] = None,
+                 function: Optional[str] = None,
+                 ) -> Union[E, None]:
+    """
+    Attempts to evaluate `value` in `enum` if value is not ``None``, otherwise returns ``None``.
+    Basically checks if a supplied enum value is valid and returns a readable error message
+    explaining the possible `enum` values if it isn't.
+
+    :param enum:
```
Comment (stuxcrystal) [inline vsutil/enums.py:61]: I think one can argue, `_` can also mean package-internal. I would keep the `_`

File: vsutil/enums.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    Enum for zimg_pixel_range_e.
+    """
+    LIMITED = 0
+    """Studio (TV) legal range, 16-235 in 8 bits."""
+    FULL = 1
+    """Full (PC) dynamic range, 0-255 in 8 bits."""
+
+
+def readable_enums(enum: Type[Enum], module: Optional[str] = None) -> str:
```
Comment (stuxcrystal) [inline vsutil/enums.py:37]: See `resolve_enum`

File: vsutil/main.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
+"""
```
Comment (stuxcrystal) [inline vsutil/main.py:1]: In the end actual bulk of the library wasn't split, wasn't it?

File: vsutil/_util.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+"""
+Private functions not meant to be exported by the main package.
+"""
+__all__ = []
```
Comment (stuxcrystal) [inline vsutil/_util.py:4]: Just leave it out.

File: vsutil/main.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+    dither = should_dither(curr_depth, bitdepth, range_in, range, clip.format.sample_type, sample_type)
+    dither_type = fallback(dither_type, enums.Dither.ERROR_DIFFUSION if dither else enums.Dither.NONE)
+
+    new_format = clip.format.replace(bits_per_sample=bitdepth, sample_type=sample_type).id
+
+    return clip.resize.Point(format=new_format, range=range, range_in=range_in, dither_type=dither_type)
+
+
+def should_dither(in_bits: int,
```
[the remark this answers is not in the document]
Comment (OrangeChannel) [reply vsutil/main.py:308]: Yea, sorry was merging/cutting a bunch of stuff here from my working directory so this is just about the file structure for now, wll revert all these changes once we can agree on something.

File: vsutil/_util.py (hunk 4 above)
Comment (OrangeChannel) [reply vsutil/_util.py:4]: Oh woops, yea was going to fill that with something but forgot I left it there.

File: vsutil/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (OrangeChannel) [reply vsutil/main.py:308]: I have some documentation stuff changed too, and the signature for resolve enum i think, I'll split those into a different PR after.

File: vsutil/__init__.py [hunk 6]
```diff
@@ -1,418 +1,15 @@
...
 
-from enum import Enum
-from functools import wraps
-from mimetypes import types_map
-from os import path
-from typing import Any, Callable, cast, List, Optional, Tuple, Type, TypeVar, Union
+# export all public function directly
+from .decorators import *
+from .enums import *
+from .main import *
```
[the remark this answers is not in the document]
Comment (OrangeChannel) [reply vsutil/__init__.py:9]: So far, from Discord, we have `std.py`, `general.py`, and `clip.py`. I think general works best here. Unless we want to further split out the functions that explicitly return a clip.

[the remark this answers is not in the document]
Comment (stuxcrystal) [reply vsutil/__init__.py:9]: I'd split it out further: https://github.com/Irrational-Encoding-Wizardry/vsutil/pull/45#discussion_r452395185

File: vsutil/_util.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+from typing import Union
+from . import enums
+
+def peak_pixel_value(bits: int, range: Union[int, enums.Range], chroma: bool) -> int:
```
[the remark this answers is not in the document]
Comment (kageru) [reply vsutil/_util.py:9]: This also isn’t supposed to be public. We could even make it an inner function of `depth`.

[the remark this answers is not in the document]
Comment (stuxcrystal) [reply vsutil/_util.py:9]: As a matter of fact: It is not public. It is a function explicitly defined in a module whose name starts with an underscore. The entire module is supposed to be internal.

Adding a `_` to the function name (if we were to leave it in this file) would just be redundant.

**Comment by OrangeChannel** [comment]
> Any reason for not splitting the module further as discussed in #38?
Wasn't sure if we'd agreed on a way to split up the main functions yet, without being too overkill and ending up with one-function modules. The current structure was what seemed to stay pretty much the same through all of the discussion. I definitely think there's at least two different types of functions in main that can be split out, ones dealing with clips and one's that don't. But I'll wait for some more feedback.

File: vsutil/_util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (OrangeChannel) [reply vsutil/_util.py:9]: Definitely could be just put inside of `scale_value(...)`, but wasn't sure if maybe this was kept separate for some specific reason, maybe a future function using this.

**Comment by stuxcrystal** [comment]
For fear of repeating myself, that was my suggestion (updated with the other comments since then)

```
vsutil
\
  __init__.py
\
  types.py
    Dither
    Range
    _readable_enums()
    _resolve_enum()
    _should_dither()

\
  func.py
    disallow_variable_format()
    disallow_variable_resolution()
    fallback()
    iterate()

\
  clips.py
    depth()
    frame2clip()
    insert_clip()
    join()
    plane()
    split()

\
  info.py
    is_image()
    get_w()
    get_y()
    get_depth()
    get_plane_size()
    get_subsampling()
```

See, no-single-function module.

**Comment by OrangeChannel** [comment]
I think we can move should_dither() into depth, and peak_pixel_value() into scale_value(), but other than that this looks good @stuxcrystal .

**Comment by OrangeChannel** [comment]
@stuxcrystal can't name the file `types.py` actually because stdlib `enum.py` imports the stdlib `types.py` and it's turned into a circular import:

```
Fatal Python error: init_import_size: Failed to import the site module
Python runtime state: initialized
Traceback (most recent call last):
  File "/usr/lib/python3.8/site.py", line 169, in addpackage
    exec(line)
  File "<string>", line 1, in <module>
  File "/home/stalled/Documents/vsutil/vsutil/types.py", line 7, in <module>
    from enum import Enum
  File "/usr/lib/python3.8/enum.py", line 2, in <module>
    from types import MappingProxyType, DynamicClassAttribute
ImportError: cannot import name 'MappingProxyType' from partially initialized module 'types' (most likely due to a circular import) (/home/stalled/Documents/vsutil/vsutil/types.py)
```

**Comment by stuxcrystal** [comment]
1. Check your sys.path
2. I personally never do relative imports.

**Comment by OrangeChannel** [comment]
I'll handle the empty docstrings and any other missing or lacking documentation in the next PR for setting up Sphinx actions in the repo, for now these can probably just be left as is.

**Pull request merged**